src/Entity/OgMembership.php

```diff
@@ -170,50 +177,47 @@ public function getGroupId() {
    * @return bool
    *   Whether or not the group is already present.
    */
-  protected function hasGroup() {
+  protected function hasGroup(): bool {
     return !empty($this->get('entity_type')->value) && !empty($this->get('entity_bundle')->value) && !empty($this->get('entity_id')->value);
   }
 
   /**
    * {@inheritdoc}
    */
-  public function getGroup() {
+  public function getGroup(): ?ContentEntityInterface {
+    assert(!empty($this->get('entity_type')->value) || !empty($this->get('entity_id')->value), new \LogicException(__METHOD__ . '() should only be called on loaded memberships, or on newly created memberships that already have the group set.'));
     $entity_type = $this->get('entity_type')->value;
     $entity_id = $this->get('entity_id')->value;
 
-    if (empty($entity_type) || empty($entity_id)) {
```

Comment:
No this case is never going to happen since the `assert()` is doing the same check and throwing an exception before this line is executed.

The only reason this should ever return `NULL` is if the entity storage deletes the group before the membership is deleted. This was happening in `OgMembershipManager::invalidateTagsOnDelete()`, which was one of the failing tests.

... And now that I am writing this I realize that this might actually be expected behaviour. An OgMembership entity _depends_ on the group. If the group is deleted then the membership _should_ be deleted alongside it.

It is possible this use case is not yet tested. I found a test case that checks that the memberships are deleted when a user is deleted (`\Drupal\Tests\og\Kernel\Entity\GetMembershipsTest::testOrphanedMembershipsDeletion()`) but could not find one for when the group is deleted.

I will investigate this case, write a test and see what happens. It might be that we can change this signature to be always returning a membership after all.